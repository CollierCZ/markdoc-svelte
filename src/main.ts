import MarkdocSource, { Config, validate } from "@markdoc/markdoc";
import yaml from "js-yaml";
import render from "./render";
import loadSchema from "./loader";
import { getComponentImports } from "./getComponents";
import getPartials from "./getPartials";

interface Options {
  extensions?: string[];
  layout?: string;
  validationLevel?: "debug" | "info" | "warning" | "error" | "critical";
  schema?: string;
  functions?: Config["functions"];
  nodes?: Config["nodes"];
  partials?: string;
  tags?: Config["tags"];
  variables?: Config["variables"];
}

interface PreprocessorReturn {
  code: string;
  data?: Record<string, unknown>;
  map?: string;
}

interface Preprocessor {
  markup: (args: {
    content: string;
    filename: string;
  }) => Promise<PreprocessorReturn | undefined>;
}

/**
 * A Svelte preprocessor for Markdoc files
 */
export const markdoc = (options: Options = {}): Preprocessor => {
  const layoutPath = options.layout;
  const schemaPath = options.schema;
  const extensions = options.extensions || [".markdown", ".md"];
  const validationLevel = options.validationLevel || "error";
  const {
    functions,
    nodes,
    partials: partialsDirectory,
    tags,
    variables,
  } = options;

  return {
    markup: async ({ content = "", filename = "" }) => {
      if (!extensions.find((extension) => filename.endsWith(extension))) return;

      const ast = MarkdocSource.parse(content);

      const isFrontmatter = Boolean(ast.attributes.frontmatter);
      const frontmatter = isFrontmatter
        ? (yaml.load(ast.attributes.frontmatter) as Record<string, unknown>)
        : {};

      const schemaFromPath = await loadSchema(schemaPath);

      const {
        partials: partialsDirectoryFromSchema,
        variables: variablesFromSchema,
        ...schemaFromPathWithoutPartials
      } = schemaFromPath;

      // Include schema parts passed as options
      // But ignore if undefined
      // Leave out partials until directory processed
      const schemaWithoutPartials = {
        ...schemaFromPathWithoutPartials,
        ...(functions && { functions }),
        ...(nodes && { nodes }),
        ...(tags && { tags }),
      };

      const markdocConfig = {
        ...schemaWithoutPartials,
        variables: { frontmatter, ...(variables || variablesFromSchema) },
        partials:
          partialsDirectory || schemaFromPath["partials"]
            ? getPartials(partialsDirectory || schemaFromPath["partials"])
            : undefined,
      };

      // Check if Markdoc is valid
      const errorLevelsMap = new Map<Options["validationLevel"], number>([
        ["debug", 0],
        ["info", 1],
        ["warning", 2],
        ["error", 3],
        ["critical", 4],
      ]);
      const errors = MarkdocSource.validate(ast, markdocConfig);
      const breakingLevel = errorLevelsMap.get(validationLevel)!;
      const areErrorsAtBreakingLevel = errors.find(
        (error) => errorLevelsMap.get(error.error.level)! >= breakingLevel
      );
      if (areErrorsAtBreakingLevel) {
        throw new Error(
          `The file at ${filename} is invalid with ${errors.length} error${errors.length > 1 ? "s" : ""}:\n- ${errors.map((error) => error.error.message).join("\n- ")}`
        );
      }

      const transformedContent = await MarkdocSource.transform(
        ast,
        markdocConfig
      );

      const svelteContent = render(transformedContent);
      const frontmatterString = isFrontmatter
        ? `<script context="module">\n` +
          `\texport const metadata = ${JSON.stringify(frontmatter)};\n` +
          `\tconst { ${Object.keys(frontmatter as Record<string, unknown>).join(
            ", "
          )} } = metadata;\n` +
          "</script>\n"
        : "";

      const componentsString = getComponentImports(
        schemaWithoutPartials,
        "/src/lib/components"
      );
      const layoutOpenString =
        layoutPath || componentsString
          ? `
<script>
  ${layoutPath ? `import Layout_DEFAULT from '${layoutPath}';` : ""}
  ${componentsString}
</script>
${
  layoutPath ? `<Layout_DEFAULT${isFrontmatter ? ` {...metadata}` : ""}>\n` : ""
}`
          : "";

      const layoutCloseString = layoutPath ? "</Layout_DEFAULT>\n" : "";

      const code =
        frontmatterString +
        layoutOpenString +
        svelteContent +
        layoutCloseString;

      return {
        code: code,
        data: frontmatter,
      };
    },
  };
};
