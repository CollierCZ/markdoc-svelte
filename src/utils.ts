import MarkdownIt from "markdown-it";

const { escapeHtml } = MarkdownIt().utils;

const BRACKET_ESCAPE_TEST_RE = /[{}]/;
const BRACKET_ESCAPE_REPLACE_RE = /[{}]/g;
const BRACKET_REPLACEMENTS = {
  "{": "&lcub;",
  "}": "&rcub;",
};

const replaceMarkdocBrackets = (characters: string): string => {
  if (characters === "{" || characters === "}") {
    return BRACKET_REPLACEMENTS[characters];
  }
  return characters;
};

export const escapeMarkdocBrackets = (
  str: string,
  includeOtherHtml = true
): string => {
  if (BRACKET_ESCAPE_TEST_RE.test(str)) {
    return str.replace(BRACKET_ESCAPE_REPLACE_RE, replaceMarkdocBrackets);
  }
  return includeOtherHtml ? escapeHtml(str) : str;
};

const BRACKET_UNESCAPE_TEST_RE = /(&lcub;|&rcub;)/;
const BRACKET_UNESCAPE_REPLACE_RE = /(&lcub;|&rcub;)/g;

const replaceEscapedBracket = (escapedBracket: string) => {
  return (
    Object.keys(BRACKET_REPLACEMENTS).find(
      // @ts-ignore
      (key) => BRACKET_REPLACEMENTS[key] === escapedBracket
    ) || ""
  );
};

export const unescapeMarkdocBrackets = (str: string) => {
  if (BRACKET_UNESCAPE_TEST_RE.test(str)) {
    return str.replace(BRACKET_UNESCAPE_REPLACE_RE, replaceEscapedBracket);
  }
  return str;
};
