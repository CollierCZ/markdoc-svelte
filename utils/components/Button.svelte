<script lang="ts">
  let text: string;
  const logMessage = () => {
    console.log(text);
  };
</script>

<button
  class="rounded-lg bg-green-100 text-green-900 py-1 px-2 text-base"
  on:click={logMessage}
>
  Click me
</button>
