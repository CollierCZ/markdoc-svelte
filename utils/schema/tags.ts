const button = {
  render: 'Button',
  attributes: {
    text: {
      type: String
    }
  }
};

const tags = {
  button: button
}

export default tags