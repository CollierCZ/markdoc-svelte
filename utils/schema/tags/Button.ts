export const button = {
  render: "Button",
  attributes: {
    text: {
      type: String,
    },
  },
};
