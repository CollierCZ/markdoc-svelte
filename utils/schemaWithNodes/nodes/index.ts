import { heading } from "../../schema/nodes/heading";

const nodes = {
  heading,
};

export default nodes;
