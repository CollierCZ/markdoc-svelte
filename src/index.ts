import { markdoc } from "./main";
import { escapeMarkdocBrackets, unescapeMarkdocBrackets } from "./utils";

export { escapeMarkdocBrackets, markdoc, unescapeMarkdocBrackets };
