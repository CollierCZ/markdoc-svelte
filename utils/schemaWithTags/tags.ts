import { button } from "../schema/tags/Button";

const tags = {
  button,
};

export default tags;
