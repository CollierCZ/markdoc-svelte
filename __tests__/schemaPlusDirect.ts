import { markdoc } from "../src/main";
import { markdocWithComponent } from "../utils/test-constants";
import nodes from "../utils/schemas/schemaWithNodes/nodes";
import { button } from "../utils/schemas/schema/tags/Button";

test(`processes correctly when nodes passed directly and tags with schema path`, async () => {
  expect(
    await markdoc({
      nodes: nodes,
      schema: "./utils/schemas/schemaWithTags",
    }).markup({
      content: markdocWithComponent,
      filename: "test.md",
    }),
  ).toMatchSnapshot();
});

test(`processes correctly when tags passed directly and nodes with schema path`, async () => {
  expect(
    await markdoc({
      tags: {
        button,
      },
      schema: "./utils/schemas/schemaWithNodes",
    }).markup({
      content: markdocWithComponent,
      filename: "test.md",
    }),
  ).toMatchSnapshot();
});
