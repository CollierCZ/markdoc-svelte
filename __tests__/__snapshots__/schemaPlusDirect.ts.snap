// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processes correctly when nodes passed directly and tags with schema path 1`] = `
{
  "code": "<article><h1 level="1" id="this-is-some-basic-markdoc">This is some basic Markdoc</h1><p>With a paragraph.</p><p>Some text <strong>in bold</strong> and <em>in italic</em>.</p><p>And a <Link href="https://example.com" rel="noopener" class="font-semibold">link</Link>.</p><h2 level="2" id="more-fancy-stuff">More fancy stuff</h2><p>Some <code>&lt;p&gt;</code> inline code.</p><p>And a code block:</p><pre data-language="javascript"><code>&lcub;&lcub;% test %&rcub;&rcub;</code></pre><p>And even a table with a nested list:</p><table><thead><tr><th>Table header 1</th><th>Table header 2</th></tr></thead><tbody><tr><td><ul><li>Row 1 Cell 1 Item 1</li><li>Row 1 Cell 1 Item 2</li></ul></td><td>Row 1 Cell 2</td></tr><tr><td>Row 2 Cell 1</td><td>Row 2 cell 2</td></tr></tbody></table><Button text="Hello world"></Button></article>",
  "data": {},
}
`;

exports[`processes correctly when tags passed directly and nodes with schema path 1`] = `
{
  "code": "<article><h1 level="1" id="this-is-some-basic-markdoc">This is some basic Markdoc</h1><p>With a paragraph.</p><p>Some text <strong>in bold</strong> and <em>in italic</em>.</p><p>And a <Link href="https://example.com" rel="noopener" class="font-semibold">link</Link>.</p><h2 level="2" id="more-fancy-stuff">More fancy stuff</h2><p>Some <code>&lt;p&gt;</code> inline code.</p><p>And a code block:</p><pre data-language="javascript"><code>&lcub;&lcub;% test %&rcub;&rcub;</code></pre><p>And even a table with a nested list:</p><table><thead><tr><th>Table header 1</th><th>Table header 2</th></tr></thead><tbody><tr><td><ul><li>Row 1 Cell 1 Item 1</li><li>Row 1 Cell 1 Item 2</li></ul></td><td>Row 1 Cell 2</td></tr><tr><td>Row 2 Cell 1</td><td>Row 2 cell 2</td></tr></tbody></table><Button text="Hello world"></Button></article>",
  "data": {},
}
`;
